File: python/797.py
"""Project Euler Problem 797: Cyclotomic Polynomials.

Let P_n(x) be the sum of all polynomials that divide x^n - 1 but not any
smaller x^k - 1. Find Σ_{n=1}^N P_n(2).

Let F_n(x) be the cyclotomic polynomial of degree n. This can be computed by
starting with an F_n(x) = x^n - 1, then for each i, dividing each i'th
element by x^i - 1.

We can compute the sum of all factors of x^n - 1 by taking each cyclotomic
polynomial factor and adding 1 (for the cases where the cyclotomic polynomial
is one of the factors and where it's not). Call this G_n(x).

P_n(x) is just G_n(x) with the polynomials that divide smaller x^k - 1 removed.
This can be computed using Inclusion Exclusion. For any term in G_n(x), it is
included in P_n(x), excluded in P_{2n}(x), and in general multiplied by µ(k)
in P_{k*n}(x). This means we just need to multiply G_n(x) by Σ_{k=1}^{N/n}
µ(k) to get the total contribution for a specific G_n(x), and then sum that
over all n.
"""

from __future__ import annotations

from typing import List


def pre_mobius(limit: int) -> List[int]:
    """Precompute Möbius function."""
    mu = [1] * (limit + 1)
    is_prime = [True] * (limit + 1)
    is_prime[0] = is_prime[1] = False

    for i in range(2, limit + 1):
        if is_prime[i]:
            for j in range(i, limit + 1, i):
                is_prime[j] = False
                if j % (i * i) == 0:
                    mu[j] = 0
                else:
                    mu[j] = -mu[j]
    return mu


def mod_inv(a: int, m: int) -> int:
    """Modular inverse."""
    return pow(a, m - 2, m)


def pow_mod(base: int, exp: int, mod: int) -> int:
    """Modular exponentiation."""
    result = 1
    base %= mod
    while exp > 0:
        if exp & 1:
            result = (result * base) % mod
        base = (base * base) % mod
        exp >>= 1
    return result


def solve() -> int:
    """Solve Problem 797."""
    N = 10**7
    M = 10**9 + 7

    mobius = pre_mobius(N)

    # Precompute Mertens function
    mertens = [0] * (N + 1)
    for i in range(1, N + 1):
        mertens[i] = mertens[i - 1] + mobius[i]

    # Compute F_n(2) = cyclotomic polynomials evaluated at 2
    F = [pow_mod(2, i, M) - 1 for i in range(N + 1)]

    # Divide by x^i - 1 for each i
    for i in range(1, N + 1):
        inv = mod_inv(F[i], M)
        for j in range(2 * i, N + 1, i):
            F[j] = (F[j] * inv) % M

    # Compute G_n(2) = product of (F_i(2) + 1) for all i|n
    G = [1] * (N + 1)
    for i in range(1, N + 1):
        for j in range(i, N + 1, i):
            G[j] = (G[j] * (F[i] + 1)) % M

    # Sum contributions
    ans = 0
    for i in range(1, N + 1):
        ans = (ans + mertens[N // i] * G[i]) % M

    return ans % M


def main() -> int:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/797.c
/*
 * Project Euler 797 - Cyclotomic Polynomials
 *
 * Let F_n(x) be the nth cyclotomic polynomial. We compute F_n(2) by sieve division.
 * G_n(2) = product of (F_i(2)+1) for all i|n is the sum of all factor-subsets of x^n-1 evaluated at 2.
 * P_n(x) = sum of polynomials dividing x^n-1 but not any smaller x^k-1.
 * By Mobius inversion, total = sum_{n=1}^N Mertens(N/n) * G_n(2).
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef long long ll;

#define N 10000000
#define MOD 1000000007LL

static ll F[N + 1];
static ll G[N + 1];
static int mu[N + 1];
static ll mertens[N + 1];
static char is_prime[N + 1];

ll pow_mod(ll base, ll exp, ll mod) {
    ll result = 1;
    base %= mod;
    if (base < 0) base += mod;
    while (exp > 0) {
        if (exp & 1) result = result * base % mod;
        base = base * base % mod;
        exp >>= 1;
    }
    return result;
}

ll mod_inv(ll a, ll m) {
    return pow_mod(a, m - 2, m);
}

int main(void) {
    /* Compute Mobius function */
    memset(is_prime, 1, sizeof(is_prime));
    is_prime[0] = is_prime[1] = 0;
    for (int i = 0; i <= N; i++) mu[i] = 1;

    for (int i = 2; i <= N; i++) {
        if (is_prime[i]) {
            for (int j = i; j <= N; j += i) {
                if (j > i) is_prime[j] = 0;
                if ((ll)(j / i) % i == 0) {
                    mu[j] = 0;
                } else {
                    mu[j] = -mu[j];
                }
            }
        }
    }

    /* Mertens function */
    mertens[0] = 0;
    for (int i = 1; i <= N; i++) {
        mertens[i] = mertens[i - 1] + mu[i];
    }

    /* Compute F_n(2) = cyclotomic polynomial evaluated at 2 */
    /* Start with F[n] = 2^n - 1, then divide by F[d] for each d|n, d<n */
    for (int i = 0; i <= N; i++) {
        F[i] = (pow_mod(2, i, MOD) - 1 + MOD) % MOD;
    }

    for (int i = 1; i <= N; i++) {
        ll inv = mod_inv(F[i], MOD);
        for (ll j = 2 * i; j <= N; j += i) {
            F[j] = F[j] * inv % MOD;
        }
    }

    /* Compute G_n(2) = product of (F_d(2)+1) for all d|n */
    for (int i = 0; i <= N; i++) G[i] = 1;
    for (int i = 1; i <= N; i++) {
        ll factor = (F[i] + 1) % MOD;
        for (ll j = i; j <= N; j += i) {
            G[j] = G[j] * factor % MOD;
        }
    }

    /* Sum contributions: ans = sum_{n=1}^N Mertens(N/n) * G_n(2) */
    ll ans = 0;
    for (int i = 1; i <= N; i++) {
        ans = (ans + (mertens[N / i] % MOD + MOD) % MOD * G[i]) % MOD;
    }

    printf("%lld\n", ans);
    return 0;
}
